### PR TITLE
Reduce Level of Nesting in UDQASTNode::eval

### PR DESCRIPTION
**Pull request by bska**
In particular, introduce a number of helper functions for evaluating expressions, well variables, group variables, unary and binary function expressions, and direct number values.  This also, I believe, fixes a latent error in one group variable case which appears not to apply the 'sign' correctly.

While here, also wrap some long lines throughout this component. This is in preparation of adding support for UDQs at the segment level.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
I'm marking this as "ready for review" now.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by blattms** [comment]
> While here, also wrap some long lines throughout this component. This is in preparation of adding support for UDQs at the segment level.

... and change a lot of the formatting. IMHO it would be nice to keep these small as they make the review a lot more time consuming and challenging.

**Review by blattms — COMMENTED** [review]
I think this is mostly good. I faintly recall thinking that some refactorings are really nice. Just forgot about them when fighting my temper due to other changes 😅

Just some comments (that might be due to my too big EGO), though.

File: src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp [hunk 1]
```diff
@@ -16,91 +16,130 @@
...
-        return UDQVarType::SCALAR;
+    if (Opm::UDQ::scalarFunc(token_type)) {
+        return Opm::UDQVarType::SCALAR;
+    }
 
-    return UDQVarType::NONE;
+    return Opm::UDQVarType::NONE;
 }
 
+} // Anonymou namespace
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp]: Thanks.  I've fixed this now.

File: src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp [hunk 2]
```diff
@@ -16,91 +16,130 @@
...
 {
+    if ((this->type == UDQTokenType::error) ||
+        (this->type == UDQTokenType::binary_op_add) ||
+        (this->type == UDQTokenType::binary_op_sub))
+    {
+        return;
+    }
+
+    throw std::invalid_argument {
+        "The one argument constructor is only available for error and end"
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp]: Leftover from #1348 (89ab99e6b).  I've updated the diagnostic message.

File: src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp [hunk 3]
```diff
@@ -16,91 +16,130 @@
...
+        this->var_type = UDQ::targetType(std::get<std::string>(this->value), this->selector);
+    }
+
+    if (this->var_type == UDQVarType::CONNECTION_VAR ||
+        this->var_type == UDQVarType::REGION_VAR ||
+        this->var_type == UDQVarType::SEGMENT_VAR ||
+        this->var_type == UDQVarType::AQUIFER_VAR ||
+        this->var_type == UDQVarType::BLOCK_VAR)
+    {
+        throw std::logic_error {
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQASTNode.cpp]: > Why not `invalid_argument` or `out_of_range`?

No reason.  I've switched to `invalid_argument`.

File: src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp [hunk 4]
```diff
@@ -74,113 +98,172 @@ namespace Opm {
...
-        this->m_definitions.insert( std::make_pair(quantity, UDQDefine(this->udq_params, quantity, report_step, location, expression)));
+        this->m_definitions.erase(quantity);
+        this->m_definitions.emplace(std::piecewise_construct,
+                                    std::forward_as_tuple(quantity),
+                                    std::forward_as_tuple(this->udq_params,
+                                                          quantity,
+                                                          report_step,
+                                                          location,
+                                                          expression));
+
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp]: > Is there a reason not to simply use `std::unordered_map::insert_or_assign()`?

Partly.  `Insert_or_assign()` requires object construction and a move, whereas `emplace()` with `piecewise_construct` and `forward_as_tuple()` will construct the final element type&mdash;in this case a `std::pair<const std::string, UDQDefine>`&mdash;in place.  That said, if `erase(key)` actually erases it's probably a wash between the two alternatives.  I've switched to using `insert_or_assign()`.

> We might want to use the returned value of erase as a hint for `std::unordered_map::emplace_hint()`.

Not possible.  `Erase(key)` returns the number of removed elements, not their position.

File: src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp [hunk 5]
```diff
@@ -193,164 +276,188 @@ namespace Opm {
...
-            return true;
-
-        if (this->m_definitions.count(keyword) > 0)
-            return true;
-
-        return false;
+    bool UDQConfig::has_keyword(const std::string& keyword) const
+    {
+        return (this->m_assignments.find(keyword) != this->m_assignments.end())
+            || (this->m_definitions.find(keyword) != this->m_definitions.end());
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp:381]: > IMHO less expressive than count but of course does the same as before.

`Find() != end()` is the canonical pre-C++20 method of checking if an element exists in the container.  It is also never more costly, and sometimes more efficient, than `count() > 0`.  In C++20 and beyond we get `unordered_map<>::contains()` which is exactly what we need here.

**Comment by bska** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (blattms) [reply src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp]: Thanks for switching.

there acually is at least one erase that return an iterator, it seems:
```
auto pos = this->m_definitions.find(quantity);
pos = this->m_definitions.erase(pos);
```

[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp]: > there actually is at least one erase that return an iterator, it seems:

That seems like a bug to me, at least if `quantity` is of the map's key type.  There are basically [two kinds of overloads](https://en.cppreference.com/w/cpp/container/unordered_map/erase) for `unordered_map<>::erase()`, one which takes an iterator and returns another iterator, and one which takes the key type and returns a count.

**Pull request merged**